learn/security/api-design-first-security-guide.md

```diff
@@ -0,0 +1,973 @@
+
+For example, if an API endpoint expects a user's age, validation rules should ensure the input is an integer and falls within a reasonable range (e.g., 0-130).
+
+Proper data validation is a critical defense against a wide range of attacks, notably **Injection flaws**. An injection attack occurs when a malicious actor sends executable code disguised as data.
+
+If the API fails to validate the input and passes it directly to a database or the operating system, that malicious code could be executed. By strictly validating that all inputs are what they are supposed to be, an API can reject malicious payloads before they can do any harm.
+
+### SQL Injection Prevention
+
+For interactions with a database, the gold standard for preventing SQL injection attacks is the use of parameterized queries, also known as prepared statements.
+
+A parameterized query forces a separation between the SQL command (the code) and the user-supplied data, making it impossible for an attacker to alter the logic of the query.
+
+**Vulnerable Code (Never Do This):**
+```python
+# DANGEROUS: Directly interpolating user input
+user_id = request.get('user_id')
+query = f"SELECT * FROM users WHERE id = {user_id}"
+cursor.execute(query)
+```
+
+**Secure Code (Use Parameterized Queries):**
+```python
+# SAFE: Using parameterized queries
+user_id = request.get('user_id')
+query = "SELECT * FROM users WHERE id = ?"
+cursor.execute(query, (user_id,))
+```
+
+### Schema-Based Validation as Security Contract
+
+OpenAPI 3.1 provides a comprehensive vocabulary for defining strict validation rules by leveraging JSON Schema Draft 2020-12. By codifying these rules directly in your API specification, validation becomes core to your API's design.
+
+**Secure Schema Example:**
+```yaml
+# openapi.yaml
+components:
+  schemas:
+    NewUser:
+      type: object
+      additionalProperties: false  # Prevent mass assignment
+      required:
+        - username
+        - email
+        - password
+        - age
+      properties:
+        username:
+          type: string
+          minLength: 4
+          maxLength: 20
+          pattern: "^[a-zA-Z0-9]+$"  # Alphanumeric only
+        email:
+          type: string
+          maxLength: 254
+          format: email
+        password:
+          type: string
+          minLength: 12
+          maxLength: 128
+        age:
+          type: integer
+          minimum: 18
+          maximum: 130
+        role:
+          type: string
+          enum: ["user", "viewer"]
+          default: "user"
+```
+
+**Redocly Governance for Validation:**
+
+Detailed rule configuration is available in the [Redocly CLI rules documentation](https://redocly.com/docs/cli/rules/).
+
+```yaml
+# redocly.yaml - Enforce security validation standards
```

Comment:
I find these multiple redocly.yaml files confusing in the way this is written. Show me the total best practice and then explain it. But don't give it to me in pieces so that if I copy something I accidentally didn't actually copy the full recommended security configuration.